File: spine-cocos2dx/src/spine-cocos2dx/Atlas.cpp
/*******************************************************************************
 * Copyright (c) 2013, Esoteric Software
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

#include <spine-cocos2dx/Atlas.h>
#include <stdexcept>

USING_NS_CC;

namespace spine {

AtlasPage::~AtlasPage () {
	CC_SAFE_RELEASE_NULL(texture);
	CC_SAFE_RELEASE_NULL(atlas);
}

//

Atlas::Atlas (const std::string &path) {
	unsigned long size;
	char *data = reinterpret_cast<char*>(CCFileUtils::sharedFileUtils()->getFileData(
			CCFileUtils::sharedFileUtils()->fullPathForFilename(path.c_str()).c_str(), "r", &size));
	if (!data) throw std::runtime_error("Error reading atlas file: " + path);
	load(data, data + size);
}

Atlas::Atlas (std::istream &input) {
	load(input);
}

Atlas::Atlas (const char *begin, const char *end) {
	load(begin, end);
}

BaseAtlasPage* Atlas::newAtlasPage (const std::string &name) {
	AtlasPage *page = new AtlasPage();
	page->texture = CCTextureCache::sharedTextureCache()->addImage(name.c_str());
	page->atlas = CCTextureAtlas::createWithTexture(page->texture, 4);
	page->atlas->retain();
	return page;
}

BaseAtlasRegion* Atlas::newAtlasRegion (BaseAtlasPage* page) {
	AtlasRegion *region = new AtlasRegion();
	region->page = reinterpret_cast<AtlasPage*>(page);
	return region;
}

AtlasRegion* Atlas::findRegion (const std::string &name) {
	return reinterpret_cast<AtlasRegion*>(BaseAtlas::findRegion(name));
}

} /* namespace spine */

File: spine-lua/Atlas.lua
-------------------------------------------------------------------------------
-- Spine Runtimes Software License v2.5
--
-- Copyright (c) 2013-2016, Esoteric Software
-- All rights reserved.
--
-- You are granted a perpetual, non-exclusive, non-sublicensable, and
-- non-transferable license to use, install, execute, and perform the Spine
-- Runtimes software and derivative works solely for personal or internal
-- use. Without the written permission of Esoteric Software (see Section 2 of
-- the Spine Software License Agreement), you may not (a) modify, translate,
-- adapt, or develop new applications using the Spine Runtimes or otherwise
-- create derivative works or improvements of the Spine Runtimes or (b) remove,
-- delete, alter, or obscure any trademarks or any copyright, trademark, patent,
-- or other intellectual property or proprietary rights notices on or in the
-- Software, including any copy thereof. Redistributions in binary or source
-- form must include this license and terms.
--
-- THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE "AS IS" AND ANY EXPRESS OR
-- IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
-- MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
-- EVENT SHALL ESOTERIC SOFTWARE BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-- SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
-- PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES, BUSINESS INTERRUPTION, OR LOSS OF
-- USE, DATA, OR PROFITS) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
-- IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
-- ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
-- POSSIBILITY OF SUCH DAMAGE.
-------------------------------------------------------------------------------

local Atlas = {}

function Atlas.parse(atlasPath, atlasBase)
	local function parseIntTuple4( l )
		local a,b,c,d = string.match( l , " ? ?%a+: ([+-]?%d+), ?([+-]?%d+), ?([+-]?%d+), ?([+-]?%d+)" )
		local a,b,c,d = tonumber( a ), tonumber( b ), tonumber( c ), tonumber( d )
		return a and b and c and d and {a, b, c ,d}
	end

	local function parseIntTuple2( l )
		local a,b = string.match( l , " ? ?%a+: ([+-]?%d+), ?([+-]?%d+)" )
		local a,b = tonumber( a ), tonumber( b )
		return a and b and {a, b}
	end

	if not atlasPath then
		error("Error: " .. atlasPath .. ".atlas" .. " doesn't exist!")
		return nil
	end

	local atlasLines = spine.utils.readFile( atlasPath, atlasBase )
	if not atlasLines then
		error("Error: " .. atlasPath .. ".atlas" .. " unable to read!")
		return nil
	end

	local pages = {}


	local it = string.gmatch(atlasLines, "(.-)\r?\n") -- iterate over lines
	for l in it do
		if #l == 0 then
			l = it()
			if l then
				local page = { name = l }
				l = it()
				page.size = parseIntTuple2( l )
				if page.size then
					l = it()
				end
				page.format = string.match( l, "%a+: (.+)" )
				page.filter = {string.match( it(), "%a+: (.+),(.+)" )}
				page.wrap = string.match( it(), "%a+: (.+)" )
				page.regions = {}
				table.insert( pages, page )
			else
				break
			end
		else
			local region = {name = l}

			region.rotate = string.match( it(), "%a+: (.+)" ) == "true"
			region.xy = parseIntTuple2( it() )
			region.size = parseIntTuple2( it() )
			l = it()
			region.splits = parseIntTuple4(l)
			if region.splits then
				l = it()
				region.pad = parseIntTuple4(l)
				if region.pad then
					l = it()
				end
			end
			region.orig = parseIntTuple2( l )
			region.offset = parseIntTuple2( it() )
			region.index = tonumber( string.match( it() , "%a+: ([+-]?%d+)" ) )

			table.insert( pages[#pages].regions, region )
		end
	end

	return pages
end

return Atlas
